torch/_inductor/codegen/halide.py

```diff
@@ -591,42 +634,350 @@ def __init__(
         self.indexing_code_dom = IndentedBuffer()
         self.needs_dom_indexing = self.inside_reduction
         self.has_reduction = self.inside_reduction
-        self.store_buffer_dimensions: Dict[str, List[sympy.Expr]] = {}
+        self.buffer_dimensions: Dict[str, List[DimensionInfo]] = {}
+        self.buffer_offsets: Dict[str, sympy.Expr] = {}
+        self.halide_vars: Dict[sympy.Symbol, sympy.Expr] = {}
+        self.index_replacements: Dict[sympy.Expr, sympy.Expr] = {}
+        self.reduction_renames: Dict[sympy.Symbol, sympy.Symbol] = {}
+        self.dom_renames: Dict[str, Dict[sympy.Symbol, sympy.Symbol]] = {}
+        self.buffer_aliases: Dict[str, List[str]] = defaultdict(list)
 
     def create_cse_var(self, name, bounds=None):
         self.body.writeline(f"{name} = hl.Func({name!r})")
         return HalideCSEVariable(name, bounds)
 
-    def codegen_iteration_ranges_entry(self, entry: IterationRangesEntry):
-        expr = self.rename_indexing(entry.expr)
-        self.indexing_code.writeline(f"{entry.name} = {self.kexpr(expr)}")
-
-        if self.has_reduction:
-            # idom includes iteration ranges of the numel of inputs
-            expr_idom = sympy_subs(
-                expr,
-                {
-                    tree.symbol(): sympy_index_symbol(f"{tree.name}_idom")
-                    for tree in self.range_trees
-                },
+    def finalize_indexing(self, indices: Sequence[sympy.Expr]):
+        """
+        Hook called right before codegen with every index that will be
+        used in the fused kernel.
+
+        This populates self.halide_vars/index_replacements/reduction_renames which is an alternate indexing
+        scheme that avoids using divide and modulus.  Instead of xindex/yindex/rindex
+        we base indexing on a larger number of vars whose product combines to those.
+
+        This function populates self.halide_vars, self.index_replacements, and self.reduction_renames
+        """
+        assert not (
+            self.index_replacements or self.halide_vars or self.reduction_renames
+        )
+        eq = V.graph.sizevars.statically_known_equals
+        lt = V.graph.sizevars.statically_known_lt
+        size_hint = functools.partial(V.graph.sizevars.size_hint, fallback=inf)
```

Comment:
Should we use an integer rather than a float for the fallback value?

Reply:
I want it to go last and there is no such thing as a max int in python

Reply:
Yeah I suppose, seems like a style preference.